### PR TITLE
Refactor all Tree related components.

### PR DESCRIPTION
**Pull request by simonguo**
## Refactor components

- [x] Refactor CheckTree
- [x] Refactor CheckTreePicker
- [x] Refactor Tree
- [x] Refactor TreePicker

## Features
- feat: add support for `searchable` on `<Tree>` and `<CheckTree>`.

## Bug Fixes
- fix(Accessibility): improve unreasonable DOM layer structure.
- fix: https://github.com/rsuite/rsuite/issues/3559
- fix: https://github.com/rsuite/rsuite/issues/3623
- fix(Tree,CheckTree): fix indent line not show in dark mode.
- fix(Tree,CheckTree): fix tree node cannot be interacted with by keyboard when virtualized.
- fix(TreePicker): fix accessibility issues.

## BREAKING CHANGE 🚨

In TreePicker and CheckTreePicker components, the following props have been deprecated and replaced with new props:

| Deprecated props | New props |
| ---------------- | --------- |
| ~menuClassName~    | popupClassName |
| ~menuStyle~        | popupStyle |
| ~renderMenu~       | renderTree |
| ~height~           | treeHeight |

### CODE REVIEW

**Comment by superman66** [comment]
# 问题记录📝

## 1、CheckTreePicker 异步加载数据未显示数据

Demo: [CheckTreePicker async ](https://rsuite-nextjs-git-refactor-tree-rsuite.vercel.app/components/check-tree-picker/#async)

截图👇

image

## 2、TreePicker 异步加载数据未显示数据


Demo: [TreePicker async ](https://rsuite-nextjs-git-refactor-tree-rsuite.vercel.app/components/tree-picker/#async)

截图👇

image


## 3、CheckTreePicker 清空数据后，再打开，会出现数据闪烁

Demo: [check-tree-picker/#basic](https://rsuite-nextjs-git-refactor-tree-rsuite.vercel.app/components/check-tree-picker/#basic)

对比了 CheckPicker 和官网上的表现，并未出现该行为。因此可以认为是重构导致的显示异常

https://github.com/rsuite/rsuite/assets/12592949/46ebcd05-bb44-481b-bada-d32c9154846a

File: src/Tree/hooks/useTreeWithChildren.ts [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+export default function useTreeWithChildren<T extends TreeNode>(
+  data: T[],
+  options: UseTreeWithChildrenOptions
+) {
+  const { valueKey, childrenKey } = options;
+  const [loadingNodeValues, setLoadingNodeValues] = useState([]);
+  const [treeData, setTreeData] = useState(data);
+
+  useEffect(() => {
+    setTreeData(treeData);
```
Comment (superman66) [inline src/Tree/hooks/useTreeWithChildren.ts:23]: 这个里面是不是写错了，没有处理 data 变更。导致异步加载时，数据无法正常显示。

```
useEffect(() => {
  setTreeData(data)  
},[data])

```

Comment (simonguo) [reply src/Tree/hooks/useTreeWithChildren.ts]: 我也发现了，修复更新了

**Comment by simonguo** [comment]
@superman66  前面提到的3 个问题已修复。

**Review by superman66 — APPROVED** [review]

**Pull request merged**